// bean-sdk/build/cli/commands/commands.js
'use strict';

var program = require('./program');
var ble = require('./ble');
var bean = require('./bean');
var install = require('./install');

function cmd(fn) {
  function _closure() {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var callback = args[args.length - 1];
    try {
      fn.apply(fn, args);
    } catch (error) {
      callback('There was an error: ' + error.message);
    }
  }
  return _closure;
}

module.exports = {
  programFirmware: cmd(program.programFirmware),
  programSketch: cmd(program.programSketch),
  startScan: cmd(ble.startScan),
  blinkLed: cmd(bean.blinkLed),
  readAccel: cmd(bean.readAccel),
  readConfig: cmd(bean.readConfig),
  readDeviceInfo: cmd(bean.readDeviceInfo),
  installBeanArduinoCore: cmd(install.installBeanArduinoCore),
  listCompiledSketches: cmd(program.listCompiledSketches),
  logSerial: cmd(bean.logSerial),
  sendSerial: cmd(bean.sendSerial),
  rename: cmd(bean.rename),
  readScratch: cmd(bean.readScratch),
  writeScratch: cmd(bean.writeScratch),
  listGATT: cmd(ble.listGATT)
};
